### PR TITLE
Fix: Destructuring defaults

### PR DESCRIPTION
**Pull request by xjamundx**
Refs #56 

- [x] not throwing errors when encountering destructured defaults
- [x] produce spec-compliant AST
- [x] all tests pass
- [x] more tests

### CODE REVIEW

**Comment by xjamundx** [comment]
Okay, finally tracked down a spec I could use:

The acorn implementation of `var {x = 10} = x` looks like:

```js
{ type: 'Program',
  start: 0,
  end: 16,
  body: 
   [ { type: 'VariableDeclaration',
       start: 0,
       end: 16,
       declarations: 
        [ { type: 'VariableDeclarator',
            start: 4,
            end: 16,
            id: 
             { type: 'ObjectPattern',
               start: 4,
               end: 12,
               properties: 
                [ { type: 'Property',
                    start: 5,
                    end: 11,
                    method: false,
                    shorthand: true,
                    computed: false,
                    key: { type: 'Identifier', start: 5, end: 6, name: 'x' },
                    kind: 'init',
                    value: 
                     { type: 'AssignmentPattern',
                       start: 5,
                       end: 11,
                       operator: '=',
                       left: { type: 'Identifier', start: 5, end: 6, name: 'x' },
                       right: { type: 'Literal', start: 9, end: 11, value: 10, raw: '10' } } } ] },
            init: { type: 'Identifier', start: 15, end: 16, name: 'x' } } ],
       kind: 'var' } ] }
```

**Comment by nzakas** [comment]
Nice!

**Comment by xjamundx** [comment]
- 2 tests with minor column number issues.
- 3 failing because we're not handling the nested cases
- 1 more failing just because...

**Comment by xjamundx** [comment]
5 failing due to missing `range` property, should be minor stuff now.

**Comment by xjamundx** [comment]
Okay fixed that now just need to add a few more tests!

**Comment by xjamundx** [comment]
Okay this should be :+1: 

Code is all mine, because nothing in Esprima to copy, so seeking a good review!

File: espree.js [hunk 1]
```diff
@@ -2560,6 +2560,26 @@ function parseObjectProperty() {
             );
         }
 
+        // destructuring defaults
+        if (allowDestructuring && match("=")) {
+            lex();
+            var value = parseAssignmentExpression();
+            var prop = markerApply(marker, astNodeFactory.createAssignmentExpression("=", id, value));
+            prop.type = astNodeTypes.AssignmentPattern;
+            var shorthand = true; // TODO: figure out the actual value
```
Comment (xjamundx) [inline espree.js]: kind of unclear why this only works when hard-coded to `true`. We added non-shorthand tests and they seem to pass as well....

File: espree.js [hunk 2]
```diff
@@ -4356,7 +4384,7 @@ function parseParam(options) {
         if (rest) {
             throwErrorTolerant(lookahead, Messages.DefaultRestParameter);
         }
-        if (!allowDefaultParams) {
+        if (!(allowDefaultParams || allowDestructuring)) {
```
Comment (nzakas) [inline espree.js]: This reads a bit funny, can you clean it up?

File: espree.js (hunk 1 above)
Comment (nzakas) [inline espree.js]: What happens when it's not hardcoded?

File: espree.js (hunk 2 above)
Comment (xjamundx) [inline espree.js]: `if (!allowDefaultParams || !allowDesctructuring)` has a slightly diff meaning and won't work.

how about:

```js
if (allowDefaultParams || allowDestructuring) {
	lex();
	def = parseAssignmentExpression();
	++options.defaultCount;
} else {
	throwUnexpected(lookahead);
}
```

**Comment by xjamundx** [comment]
Okay give this one a lookover if you can. I figured out why shorthand was required in that block. Longform is a different code path and probably already worked :)

File: espree.js (hunk 2 above)
Comment (nzakas) [inline espree.js]: :+1:

**Comment by nzakas** [comment]
Lgtm. Good to merge?

**Comment by xjamundx** [comment]
Yeah I think it's :+1:

**Pull request merged**